File: app/src/main/java/ru/mammoth70/wherearetheynow/YandexActivity.java
package ru.mammoth70.wherearetheynow;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.PointF;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.yandex.mapkit.MapKitFactory;
import com.yandex.mapkit.geometry.Circle;
import com.yandex.mapkit.map.CameraListener;
import com.yandex.mapkit.map.CameraUpdateReason;
import com.yandex.mapkit.map.CircleMapObject;
import com.yandex.mapkit.map.IconStyle;
import com.yandex.mapkit.map.Map;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.MapWindow;
import com.yandex.mapkit.map.PlacemarkMapObject;
import com.yandex.mapkit.map.SizeChangedListener;
import com.yandex.mapkit.map.TextStyle;
import com.yandex.mapkit.mapview.MapView;
import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.CameraPosition;
import com.yandex.runtime.image.ImageProvider;

import java.util.ArrayList;
import java.util.Objects;

public class YandexActivity extends LocationActivity implements CameraListener, SizeChangedListener {
    // Activity выводит yandex-карту с геолокацией, переданной через intent.

    private MapView mapView;
    private float mapZoom;
    private float mapTilt;
    private Map map;
    private FloatingActionButton fabNord;
    private FloatingActionButton fabZoomIn;
    private FloatingActionButton fabZoomOut;
    private FloatingActionButton fab2D3D;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        // Метод вызывается при создании Activity.
        // Из intent получается координаты и выводится карта с метками.
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        MapKitFactory.initialize(this);
        setContentView(R.layout.activity_yandex);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.yandex),
                (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        createFrameTitle(this);
        fabNord = findViewById(R.id.floatingActionButtonMapNord);
        fabZoomIn = findViewById(R.id.floatingActionButtonMapZoomIn);
        fabZoomOut = findViewById(R.id.floatingActionButtonMapZoomOut);
        fab2D3D = findViewById(R.id.floatingActionButtonMapTilt);

        mapView = findViewById(R.id.yandexview);
        mapView.getMapWindow().getMap().addCameraListener(this);
        map = mapView.getMapWindow().getMap();
        int currentNightMode = getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;
        switch (currentNightMode) {
            case Configuration.UI_MODE_NIGHT_NO:
                // ночная тема не активна, используется светлая тема
                map.setNightModeEnabled(false);
                break;
            case Configuration.UI_MODE_NIGHT_YES:
                // ночная тема активна, и она используется
                map.setNightModeEnabled(true);
                break;
        }
        MapObjectCollection mapObjectCollection = mapView.getMapWindow().getMap().getMapObjects().addCollection();

        // Стартовую точку привязываем к данным, переданным через intent.
        mapZoom = intent.getFloatExtra(Util.INTENT_EXTRA_MAP_ZOOM, MapUtil.MAP_ZOOM_DEFAULT);
        mapTilt = intent.getFloatExtra(Util.INTENT_EXTRA_MAP_TILT, MapUtil.MAP_TILT_DEFAULT);
        boolean mapCircle = intent.getBooleanExtra(Util.INTENT_EXTRA_MAP_CIRCLE, MapUtil.MAP_CIRCLE_DEFAULT);

        ArrayList<Point> points = new ArrayList<>();
        ArrayList<ImageProvider> imageProviders = new ArrayList<>();
        ArrayList<PlacemarkMapObject> placemarkMapObjects = new ArrayList<>();
        ArrayList<String> markerUserData = new ArrayList<>();
        ArrayList<CircleMapObject> circleMapObjects = new ArrayList<>();

        // Добавляем все маркеры. Цикл по списку разрешённых телефонов.
        for (String key : Util.phones) {
            if (Util.phone2record.containsKey(key)) {
                PointRecord value = Util.phone2record.get(key);
                if (value != null) {
                    points.add(new Point(value.latitude, value.longitude));
                    imageProviders.add(ImageProvider.fromBitmap
                            (getBitmapFromColor(Util.phone2color.get(key))));
                    placemarkMapObjects.add(mapObjectCollection.addPlacemark());
                    if (mapCircle) {
                        circleMapObjects.add(mapObjectCollection.addCircle(
                                new Circle(new Point(value.latitude, value.longitude), 75f)));
                    }
                    markerUserData.add(key);
                }
            }
        }

        // Стиль иконки маркера.
        IconStyle iconStyle = new IconStyle();
        iconStyle.setAnchor(new PointF(0.5f,1f));

        // Стиль текста над маркером.
        TextStyle textStyle = new TextStyle();
        textStyle.setSize(12);
        textStyle.setColor(getColor(R.color.md_theme_onBackground));
        textStyle.setOutlineColor(getColor(R.color.md_theme_background));
        textStyle.setOutlineWidth(4);
        textStyle.setPlacement(TextStyle.Placement.TOP);

        // Настраиваем свойства объектов-маркеров на карте.
        for (int i = 0; i < points.size(); i++) {
            placemarkMapObjects.get(i).setGeometry(points.get(i));
            placemarkMapObjects.get(i).setIcon(imageProviders.get(i),iconStyle);
            placemarkMapObjects.get(i).setUserData(markerUserData.get(i));
            placemarkMapObjects.get(i).addTapListener(mapObjectTapListener);
            if (mapCircle) {
                circleMapObjects.get(i).setStrokeColor(Color.parseColor(Util.phone2color.get(
                        (String) placemarkMapObjects.get(i).getUserData())));
                circleMapObjects.get(i).setStrokeWidth(1.f);
                circleMapObjects.get(i).setFillColor(Color.parseColor(AppColors.getColorAlpha(
                        Util.phone2color.get((String) placemarkMapObjects.get(i).getUserData()))));
                placemarkMapObjects.get(i).setText(Objects.requireNonNull
                        (Util.phone2name.get(markerUserData.get(i))),textStyle);
            }
        }

        reloadMapFromPoint(this, startRecord);
    }

    @Override
    protected void reloadMapFromPoint(Context context, PointRecord rec) {
        // Метод перестраивает карту по PointRecord.
        start2D3D();
        Point point = new Point(rec.latitude, rec.longitude);
        CameraPosition cameraPosition = new CameraPosition(point, mapZoom, 0f, mapTilt);
        map.move(cameraPosition);
    }

    private final MapObjectTapListener mapObjectTapListener = (mapObject, point) -> {
        // Метод, отвечающий за тапы по различным объектам на карте.
        String phone = (String)mapObject.getUserData();
        String name = Util.phone2name.get(phone);
        PointRecord rec = Util.phone2record.get(phone);
        if (rec != null) {
            CameraPosition currCameraPosition = map.getCameraPosition();
            Point newPoint = new Point(rec.latitude, rec.longitude);
            float newZoom = currCameraPosition.getZoom();
            if (newZoom < mapZoom) {
                newZoom = mapZoom;
            }
            tvName.setText(Util.phone2name.get(phone));
            tvDateTime.setText(MapUtil.timePassed(
                    Objects.requireNonNull(Util.phone2record.get(phone)).datetime, this));
            CameraPosition newCameraPosition = new CameraPosition(
                    newPoint,
                    newZoom,
                    0,
                    currCameraPosition.getTilt());
            map.move(newCameraPosition);
            /*
            String message = name + "\n" +
                    String.format(Locale.US, PointRecord.FORMAT_POINT,
                            point.getLatitude(), point.getLongitude());
            Toast.makeText(getApplicationContext(), message, Toast.LENGTH_SHORT).show();
            */
        }
        return true;
    };

    private Bitmap createBitmapFromVector(int art) {
        // Функция преобразовывает векторное изображение в bitmap.
        Drawable drawable = ContextCompat.getDrawable(this, art);
        assert drawable != null;
        Bitmap bitmap = Bitmap.createBitmap(
                drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(),
                Bitmap.Config.ARGB_8888
        );
        Canvas canvas =  new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }

    private Bitmap getBitmapFromColor(String color) {
        // Метод возвращает маркер заданного цвета.
        return createBitmapFromVector(AppColors.getColorMarkerSmall(color));
    }

    protected void onStart() {
        // Метод вызывается перед тем, как Activity будет видно пользователю.
        super.onStart();
        MapKitFactory.getInstance().onStart();
        mapView.onStart();
    }

    protected void onStop() {
        // Метод вызывается, когда Activity становится не видно пользователю.
        mapView.onStop();
        MapKitFactory.getInstance().onStop();
        super.onStop();
    }

    public void onMapNordClicked(View view) {
        // Метод поворачивает карту в положение север сверху.
        CameraPosition currCameraPosition = map.getCameraPosition();
        CameraPosition newCameraPosition = new CameraPosition(
                currCameraPosition.getTarget(),
                currCameraPosition.getZoom(),
                0,
                currCameraPosition.getTilt());
        map.move(newCameraPosition);
    }

    public void onMapZoomIn(View view) {
        // Метод приближает объекты на карте.
        CameraPosition currCameraPosition = map.getCameraPosition();
        float newZoom = currCameraPosition.getZoom() + 1f;
        if (newZoom > map.getCameraBounds().getMaxZoom()) {
            newZoom = map.getCameraBounds().getMaxZoom();
        }
        CameraPosition newCameraPosition = new CameraPosition(currCameraPosition.getTarget(),
                newZoom,
                currCameraPosition.getAzimuth(),
                currCameraPosition.getTilt());
        map.move(newCameraPosition);
    }

    public void onMapZoomOut(View view) {
        // Метод отдаляет от объектов на карте.
        CameraPosition currCameraPosition = map.getCameraPosition();
        float newZoom = currCameraPosition.getZoom() - 1f;
        if (newZoom < map.getCameraBounds().getMinZoom()) {
            newZoom = map.getCameraBounds().getMinZoom();
        }
        CameraPosition newCameraPosition = new CameraPosition(
                currCameraPosition.getTarget(),
                newZoom,
                currCameraPosition.getAzimuth(),
                currCameraPosition.getTilt());
        map.move(newCameraPosition);
    }

    private void start2D3D() {
        // Метод устанавливает в начальное состояние FAB Tilt.
        if (mapTilt == 0f) {
            fab2D3D.setContentDescription(getString(R.string.map3d));
            fab2D3D.setImageResource(R.drawable.ic_action_3d);
        } else {
            fab2D3D.setContentDescription(getString(R.string.map2d));
            fab2D3D.setImageResource(R.drawable.ic_action_2d);
        }
    }

    public void setfabStatus() {
        // Метод устанавливает в разрешенное состояние FAB-ы.
        CameraPosition currCameraPosition = map.getCameraPosition();
        float zoom = currCameraPosition.getZoom();
        if (zoom <= map.getCameraBounds().getMinZoom()) {
            fabZoomOut.hide();
            fabZoomIn.show();
        } else if (zoom >= map.getCameraBounds().getMaxZoom()) {
            fabZoomIn.hide();
            fabZoomOut.show();
        } else {
            fabZoomIn.show();
            fabZoomOut.show();
        }
        if (currCameraPosition.getTilt() == 0f) {
            fab2D3D.setContentDescription(getString(R.string.map3d));
            fab2D3D.setImageResource(R.drawable.ic_action_3d);
        } else {
            fab2D3D.setContentDescription(getString(R.string.map2d));
            fab2D3D.setImageResource(R.drawable.ic_action_2d);
        }
        if (currCameraPosition.getAzimuth() == 0f) {
            fabNord.hide();
        } else {
            fabNord.show();
        }

    }

    public void onMap2D3D(View view) {
        // Метод меняет режим карты 2D / 3D.
        CameraPosition currCameraPosition = map.getCameraPosition();
        float currTilt = currCameraPosition.getTilt();
        float newTilt;
        if (currTilt == 0f) {
            if (mapTilt == 0f) {
                newTilt = MapUtil.MAP_TILT_DEFAULT;
            } else {
                newTilt = mapTilt;
            }
        } else {
            newTilt = 0f;
        }
        CameraPosition newCameraPosition = new CameraPosition(
                currCameraPosition.getTarget(),
                currCameraPosition.getZoom(),
                currCameraPosition.getAzimuth(),
                newTilt);
        map.move(newCameraPosition);
    }

    @Override
    public void onCameraPositionChanged(@NonNull Map map, @NonNull CameraPosition cameraPosition,
                                        @NonNull CameraUpdateReason cameraUpdateReason,
                                        boolean finished) {
        // Интерфейс для подписки на изменение положения камеры.
        if (finished) {
            setfabStatus();
        }
    }

    @Override
    public void onMapWindowSizeChanged(@NonNull MapWindow mapWindow,
                                       int newWidth,
                                       int newHeight) {
        // Интерфейс для подписки на изменение размеров карты.
        setfabStatus();
    }

}

File: app/src/main/java/ru/mammoth70/wherearetheynow/YandexActivity.kt
package ru.mammoth70.wherearetheynow

import android.content.Context
import android.content.res.Configuration
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.PointF
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.graphics.toColorInt
import androidx.core.graphics.createBitmap
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.yandex.mapkit.MapKitFactory
import com.yandex.mapkit.geometry.Circle
import com.yandex.mapkit.geometry.Point
import com.yandex.mapkit.map.CameraListener
import com.yandex.mapkit.map.CameraPosition
import com.yandex.mapkit.map.CameraUpdateReason
import com.yandex.mapkit.map.CircleMapObject
import com.yandex.mapkit.map.IconStyle
import com.yandex.mapkit.map.Map
import com.yandex.mapkit.map.MapObjectTapListener
import com.yandex.mapkit.map.MapWindow
import com.yandex.mapkit.map.PlacemarkMapObject
import com.yandex.mapkit.map.SizeChangedListener
import com.yandex.mapkit.map.TextStyle
import com.yandex.mapkit.mapview.MapView
import com.yandex.runtime.image.ImageProvider
import java.util.Locale

class YandexActivity : LocationActivity(), CameraListener, SizeChangedListener {
    // Activity выводит yandex-карту с геолокацией, переданной через intent.

    override val idLayout = R.layout.activity_yandex
    override val idActivity = R.id.frameYandexActivity

    private val mapView: MapView by lazy { findViewById(R.id.yandexview) }
    private val map: Map by lazy { mapView.mapWindow.map }
    private val fabNord: FloatingActionButton by lazy {
        findViewById(R.id.floatingActionButtonMapNord) }
    private val fabZoomIn: FloatingActionButton by lazy {
        findViewById(R.id.floatingActionButtonMapZoomIn) }
    private val fabZoomOut: FloatingActionButton by lazy {
        findViewById(R.id.floatingActionButtonMapZoomOut) }
    private val fab2D3D: FloatingActionButton by lazy {
        findViewById(R.id.floatingActionButtonMapTilt) }


    override fun initMap(context: Context) {
        // Функция делает начальную настройку карты.

        mapView.mapWindow.map.addCameraListener(this)
        when (getResources().configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) {
            Configuration.UI_MODE_NIGHT_NO ->
                // ночная тема не активна, используется светлая тема
                map.isNightModeEnabled = false

            Configuration.UI_MODE_NIGHT_YES ->
                // ночная тема активна, и она используется
                map.isNightModeEnabled = true
        }

        val mapObjectCollection = mapView.mapWindow.map.mapObjects.addCollection()
        val points = ArrayList<Point?>()
        val imageProviders = ArrayList<ImageProvider?>()
        val placemarkMapObjects = ArrayList<PlacemarkMapObject?>()
        val circleMapObjects = ArrayList<CircleMapObject?>()

        fabNord.setOnClickListener { _ ->
            onMapNord()
        }
        fabZoomIn.setOnClickListener { _ ->
            onMapZoomIn()
        }
        fabZoomOut.setOnClickListener { _ ->
            onMapZoomOut()
        }
        fab2D3D.setOnClickListener { _ ->
            onMap2D3D()
        }

        // Добавляем все метки. Цикл по списку разрешённых телефонов.
        DataRepository.users
            .filter { it.lastRecord != null }
            .forEach { user ->
                val point = user.lastRecord!!
                val mapPoint = Point(point.latitude, point.longitude)
                val userColor = user.color

                points.add(mapPoint)

                val iconProvider = ImageProvider.fromBitmap(getBitmapFromColor(userColor))
                imageProviders.add(iconProvider)

                placemarkMapObjects.add(
                    mapObjectCollection.addPlacemark().apply {
                        geometry = mapPoint
                        setIcon(iconProvider, IconStyle().apply { anchor = PointF(0.5f, 1f) })
                        userData = user.phone
                        addTapListener(mapObjectTapListener)
                        setText(user.name, markTextStyle) // Имя берем из объекта
                    }
                )

                if (SettingsManager.selectedMapCircle) {
                    circleMapObjects.add(
                        mapObjectCollection.addCircle(
                            Circle(mapPoint, SettingsManager.selectedMapCircleRadius)
                        ).apply {
                            strokeColor = userColor.toColorInt()
                            strokeWidth = 1f
                            fillColor = PinColors.getColorAlpha(userColor)
                        }
                    )
                }

            }
    }


    override fun reloadMapFromPoint(context: Context, rec: PointRecord) {
        // Функция передвигает карту на PointRecord.

        map.move(CameraPosition(Point(rec.latitude, rec.longitude),
            SettingsManager.selectedMapZoom, 0f, SettingsManager.selectedMapTilt))
    }


    private val markTextStyle: TextStyle
        get() {
            // Геттер настраивает стиль текста над меткой.

            return TextStyle().apply {
                size = 12f
                color = getThemeColor(R.attr.colorOnBackground)
                outlineColor = getThemeColor(android.R.attr.colorBackground)
                outlineWidth = 4f
                placement = TextStyle.Placement.TOP
            }
        }


    private val mapObjectTapListener =
        MapObjectTapListener { mapObject, _ ->
            // Обработчик, отвечающий за тапы по различным объектам на карте.

            val phone = mapObject.userData as? String ?: return@MapObjectTapListener true
            val user = DataRepository.getUser(phone) ?: return@MapObjectTapListener true

            user.lastRecord?.let { rec ->
                val currentZoom = map.cameraPosition.zoom
                val newZoom = if (currentZoom < SettingsManager.selectedMapZoom) {
                    SettingsManager.selectedMapZoom
                } else {
                    currentZoom
                }

                topAppBar.setTitle(user.name)
                topAppBar.setSubtitle(timePassed(rec.dateTime, this))

                map.move(
                    CameraPosition(
                        Point(rec.latitude, rec.longitude),
                        newZoom,
                        map.cameraPosition.azimuth,
                        map.cameraPosition.tilt
                    )
                )
                val message = "${user.name}\n${String.format(Locale.US, PointRecord.FORMAT_POINT, rec.latitude, rec.longitude)}"
                Toast.makeText(applicationContext, message, Toast.LENGTH_SHORT).show()
            }
            true
        }


    private fun createBitmapFromVector(art: Int): Bitmap? {
        // Функция преобразовывает векторное изображение в bitmap.

        val drawable = ContextCompat.getDrawable(this, art) ?: return null
        val bitmap = createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight)
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }


    private fun getBitmapFromColor(color: String?): Bitmap? {
        // Функция возвращает метку заданного цвета.

        return createBitmapFromVector(PinColors.getPin(color))
    }


    override fun onStart() {
        // Функция вызывается перед тем, как Activity будет видно пользователю.

        super.onStart()
        MapKitFactory.getInstance().onStart()
        mapView.onStart()
    }


    override fun onStop() {
        // Функция вызывается, когда Activity становится не видно пользователю.

        mapView.onStop()
        MapKitFactory.getInstance().onStop()
        super.onStop()
    }


    private fun onMap2D3D() {
        // Обработчик кнопки FAB "2D/3D".
        // Функция меняет режим карты 2D / 3D.

        val newTilt = if (map.cameraPosition.tilt == 0f) {
            if (SettingsManager.selectedMapTilt == 0f) {
                MAP_TILT_DEFAULT
            } else {
                SettingsManager.selectedMapTilt
            }
        } else {
            0f
        }
        map.move(CameraPosition(
            map.cameraPosition.target,
            map.cameraPosition.zoom,
            map.cameraPosition.azimuth,
            newTilt
        ))
    }


    private fun onMapNord() {
        // Обработчик кнопки FAB "На север".
        // Функция поворачивает карту в положение север сверху.

        map.move(CameraPosition(
            map.cameraPosition.target,
            map.cameraPosition.zoom,
            0f,
            map.cameraPosition.tilt
        ))
    }


    private fun onMapZoomIn() {
        // Обработчик кнопки FAB "Zoom In".
        // Функция приближает объекты на карте.

        val newZoom = if ((map.cameraPosition.zoom + 1f) > map.cameraBounds.maxZoom) {
            map.cameraBounds.maxZoom
        } else {
            map.cameraPosition.zoom + 1f
        }
        map.move(CameraPosition(
            map.cameraPosition.target,
            newZoom,
            map.cameraPosition.azimuth,
            map.cameraPosition.tilt
        ))
    }


    private fun onMapZoomOut() {
        // Обработчик кнопки FAB "Zoom Out".
        // Функция отдаляет объекты на карте.

        val newZoom = if ((map.cameraPosition.zoom - 1f) < map.cameraBounds.minZoom) {
            map.cameraBounds.minZoom
        } else {
            map.cameraPosition.zoom - 1f
        }
        map.move(CameraPosition(
            map.cameraPosition.target,
            newZoom,
            map.cameraPosition.azimuth,
            map.cameraPosition.tilt
        ))
    }


    private fun setFabStatus() {
        // Функция устанавливает в правильное состояние все FAB'ы.

        val zoom = map.cameraPosition.zoom
        when
            { (zoom <= map.cameraBounds.minZoom) -> {
                fabZoomOut.hide()
                fabZoomIn.show()
            } (zoom >= map.cameraBounds.maxZoom) -> {
                fabZoomIn.hide()
                fabZoomOut.show()
            } else -> {
                fabZoomIn.show()
                fabZoomOut.show()
            }
        }
        if (map.cameraPosition.tilt == 0f) {
            fab2D3D.setContentDescription(getString(R.string.map3d))
            fab2D3D.setImageResource(R.drawable.ic_action_3d)
        } else {
            fab2D3D.setContentDescription(getString(R.string.map2d))
            fab2D3D.setImageResource(R.drawable.ic_action_2d)
        }
        if (map.cameraPosition.azimuth == 0f) {
            fabNord.hide()
        } else {
            fabNord.show()
        }
    }


    override fun onCameraPositionChanged(
        map: Map, cameraPosition: CameraPosition,
        cameraUpdateReason: CameraUpdateReason,
        finished: Boolean
    ) {
        // Интерфейс для подписки на изменение положения камеры.
        if (finished) {
            setFabStatus()
        }
    }


    override fun onMapWindowSizeChanged(
        mapWindow: MapWindow,
        newWidth: Int,
        newHeight: Int
    ) {
        // Интерфейс для подписки на изменение размеров карты.
        setFabStatus()
    }

}